bootstrap.py

```diff
@@ -77,45 +80,23 @@ def region():
 
 def resource_tags():
     """ Returns a dictionary of all resource tags for the current instance """
-    import boto.ec2
-    api = boto.ec2.connect_to_region(region())
-    tags = api.get_all_tags(filters={'resource-id': instance_id()})
-    return {tag.name: tag.value for tag in tags}
-
-
-def security_groups():
-    """ Returns a list of sercurity groups for the current instance """
-    return instance_metadata('security-groups').split('\n')
-
-
-def infer_tags(security_group):
-    """ Attempts to infer tags from a security group name """
-    import re
-    matches = re.search(r'(?P<Project>[\w-]+)-(?P<Role>\w+)$', security_group)
-    return matches.groupdict()
-
-
-def implicit_tags():
-    """ Returns a list of tags inferred from security groups """
-    return [infer_tags(name) for name in security_groups()]
-
-
-def discover(trait):
-    """ Tries to find a trait in tags, makes a reasonable guess if it fails """
-    if trait in resource_tags():
-        return [resource_tags()[trait]]
-    else:
-        return [implicit_tags()[trait]]
+    try:
+        import boto.ec2
+        api = boto.ec2.connect_to_region(region())
+        tags = api.get_all_tags(filters={'resource-id': instance_id()})
+        return {tag.name: tag.value for tag in tags}
+    except:
+        pass
 
 
 def project_path():
     """ Returns the forge path for the discovered project """
-    return discover('Project')[0] + '/'
```

Comment:
This is a case where `discover()` could sanely be replaced with `detect()`, if the latter called the former. I'd say you're on the right path with this one.